res/sections/04-GameBank.tex

```diff
@@ -0,0 +1,189 @@
+transaction communications.
+When all the conditions to start the match are met, the host has the 
+possibility start it, making every connected device to display the dashboard 
+activity.
+
+The dashboard activity is subdivided in two tabs: one for transfer money and 
+the other one to see the transaction log. In the first tab, users can send 
+money to each other or to themselves (impresonating the bank). In the second 
+tab, there is the transaction log, that is updated every time someone creates a 
+new transaction. In this log, players can see all the transactions. In this 
+way, they can check if someone is cheating. \todo{Here we should put a 
+screenshot of the application}
+
+\subsection{Architecture}
+
+GameBank architecture is composed of two layers \todo{Insert figure here}: one 
+for the communications/networking and the other one for the game logic. With 
+this architecture data trasmission and game logic are highly decoupled, so that 
+is easy to make changes to one of the two parts. In particular, this could lead 
+to a simpler re-utilization of the network layer in other applications.
+
+The whole application is built using the event architecture. Every change in 
+the application state is interpreted as a new event, that has a particular tag 
+and is captured from subscribers along the application. While this seems to 
+make the flow difficult to understand and cubersome, it actually allows to keep 
+the code very decoupled.
+
+Regarding the network layer, it is built in a way that host and client 
+operations are as similar as possible. The only different parts are the way 
+clients and hosts get the MAC address of the receiver. In particular, the host 
+has to keep in memory all the clients addresses, while the clients need to 
+keep in memory only the host address. A particular object, \texttt{BTBundle} is 
+used as a wrapper where other layers can put data to send.
+
+\begin{figure}[t]
+ \centering
+ \includegraphics[scale=0.4]{GameBank}
+ \caption{GameBank architecture}
+ \label{fig:gbArchitecture}
+\end{figure}
+
+The game layer simply takes the events coming from the network transmissions, 
+and store the data in a Realm database\footnote{\url{https://realm.io/}}, that 
+is based on events too. Finally, the views get updated with the new data stored 
+in the database.
+
+\subsection{Game creation}
+
+In the section above we briefly talked how a match is created: here we are 
+going to describe in the details the procedure we have designed to connect the 
+Bluetooth devices and to start a match.
+
+When a user opens a match, in the network layer a Bluetooth socket is opened 
+and set up for an incoming connection. When a client connects to the host, first 
+it saves its UUID, then it perform a \textit{rendezvous}: in other words a 
+\texttt{BTBundle} with all the client information is sent (including player 
+username and profile picture) to the host. This is the slower phase of the 
```

Comment:
.. to host, first it save its UUID, its username and its profile picture into a BTBundle (a rendezvous) then it send it to host.

Cancellerei tutto da first ... a the host.

Reply:
Eh no cavoli, stiamo spiegando come funziona il nostro mini protocollo in un esame di reti wireless. Secondo me dovrebbe rimanere invece :sweat_smile:

Reply:
Intendevo sostituire con quello che ho scritto io

Reply:
Annn :sweat_smile: